README
NAME
    MooseX::AbstractMethod - Declare methods requirements that must be
    satisfied

VERSION
    version 0.003

SYNOPSIS
        use Moose;
        use MooseX::Abstract;

        requires 'bar';

        # synonm to 'requires'
        abstract 'foo';

DESCRIPTION
    This extensions allows classes to flag certain methods as being required
    to be implemented by a subclass, much as a Moose::Role does with
    'requires'.

USAGE
    As in the synopsis, simply mark certain methods as being required by
    subclasses by passing their names to "abstract" or "requires". This will
    cause a method of the same name to be installed in the class that will
    die horribly if it's ever called. Additionally, when a class is made
    immutable, all of its methods are checked to see if they're marked as
    abstract; if any abstract methods exists that were not created in the
    current class, we die horribly.

    Checking for method satisfaction on make_immutable isn't perfect, but
    AFAICT it's the most reasonable approach possible at the moment.
    (Corrections welcome.)

NEW SUGAR
  abstract
    abstract() allows one to declare a method dependency that must be
    satisfied by a subclass before it is invoked, and before the subclass is
    made immutable.

        abstract 'method_name_that_must_be_satisfied';

  requires
    requires() is a synonym for abstract() and works in the way you'd
    expect.

SEE ALSO
BUGS
    All complex software has bugs lurking in it, and this module is no
    exception.

    Bugs, feature requests and pull requests through GitHub are most
    welcome; our page and repo (same URI):

        https://github.com/RsrchBoy/moosex-abstractmethod

AUTHOR
    Chris Weyl <cweyl@alumni.drew.edu>

COPYRIGHT AND LICENSE
    This software is Copyright (c) 2011 by Chris Weyl.

    This is free software, licensed under:

      The GNU Lesser General Public License, Version 2.1, February 1999



Changes
Revision history for MooseX-AbstractMethod

0.003     2012-01-10 12:01:16 America/Los_Angeles
- don't mess with roles (that is, have our sugar just hand off to
  Moose::Role::requires() if we happen to be use'd from a role

0.002     2011-11-15 14:43:39 America/Los_Angeles
- documentation updates; nothing else

0.001     2011-09-16 22:09:35 America/Los_Angeles


lib/MooseX/AbstractMethod.pm
#
# This file is part of MooseX-AbstractMethod
#
# This software is Copyright (c) 2011 by Chris Weyl.
#
# This is free software, licensed under:
#
#   The GNU Lesser General Public License, Version 2.1, February 1999
#
package MooseX::AbstractMethod;
{
  $MooseX::AbstractMethod::VERSION = '0.003';
}

# ABSTRACT: Declare methods requirements that must be satisfied

use Moose 0.94 ();
use namespace::autoclean;
use Moose::Exporter;
use Moose::Util::MetaRole;

# debugging...
#use Smart::Comments;

{
    package MooseX::AbstractMethod::Trait::Class;
{
  $MooseX::AbstractMethod::Trait::Class::VERSION = '0.003';
}
    use Moose::Role;
    use namespace::autoclean;
    use Moose::Util qw{ does_role english_list };

    sub abstract_method_metaclass {
        my $self = shift @_;

        return Moose::Meta::Class
            ->create_anon_class(
                # XXX should this just be Moose::Meta::Method as the superclass??
                superclasses => [ $self->method_metaclass           ],
                roles        => [ 'MooseX::AbstractMethod::Trait::Method' ],
                cache        => 1,
            )
            ->name
            ;
    }

    sub add_abstract_method {
        my ($self, @names) = @_;

        my $abstract_meta = $self->abstract_method_metaclass;

        for my $name (@names) {

            my $method = $abstract_meta->wrap(
                sub { Moose->throw_error("Method $name is not implemented!") },
                name         => $name,
                package_name => $self->name,
            );

            $self->add_method($name => $method);
        }

        return;
    }

    # blow up if we're being asked to become immutable, we're a subclasss,
    # and we don't implement an abstract method

    before make_immutable => sub {
        my $self = shift @_;

        my @abstract_methods = sort
            map  { $_->name . ' (from ' . $_->original_package_name . ')' }
            grep { $_->original_package_name ne $self->name               }
            grep { does_role($_, 'MooseX::AbstractMethod::Trait::Method') }
            $self->get_all_methods
            ;

        Moose->throw_error(
            'These abstract methods have not been implemented in '
            . $self->name . ': '
            . english_list(@abstract_methods)
        ) if @abstract_methods;

        return;
    };

}
{
    package MooseX::AbstractMethod::Trait::Method;
{
  $MooseX::AbstractMethod::Trait::Method::VERSION = '0.003';
}
    use Moose::Role;
    use namespace::autoclean;

    # well, hello there handsome :)
}

sub abstract { _abstract(@_) }
sub requires { _abstract(@_) }

sub _abstract {

    ### meta isa: ref $_[0]
    return shift->add_abstract_method(@_)
        unless $_[0]->isa('Moose::Meta::Role');

    goto \&Moose::Role::requires;
}

Moose::Exporter->setup_import_methods(
    with_meta => [ qw{ abstract requires } ],

    trait_aliases => [
        [ 'MooseX::AbstractMethod::Trait::Method' => 'AbstractMethod' ],
    ],
    class_metaroles => {
        class => [ 'MooseX::AbstractMethod::Trait::Class' ],
    },
);

!!42;



=pod

=head1 NAME

MooseX::AbstractMethod - Declare methods requirements that must be satisfied

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Moose;
    use MooseX::Abstract;

    requires 'bar';

    # synonm to 'requires'
    abstract 'foo';

=head1 DESCRIPTION

This extensions allows classes to flag certain methods as being required to be
implemented by a subclass, much as a L<Moose::Role> does with 'requires'.

=head1 USAGE

As in the synopsis, simply mark certain methods as being required by
subclasses by passing their names to "abstract" or "requires".  This will
cause a method of the same name to be installed in the class that will die
horribly if it's ever called.  Additionally, when a class is made immutable,
all of its methods are checked to see if they're marked as abstract; if any
abstract methods exists that were not created in the current class, we die
horribly.

Checking for method satisfaction on make_immutable isn't perfect, but AFAICT
it's the most reasonable approach possible at the moment.  (Corrections
welcome.)

=head1 NEW SUGAR

=head2 abstract

abstract() allows one to declare a method dependency that must be satisfied by a
subclass before it is invoked, and before the subclass is made immutable.

    abstract 'method_name_that_must_be_satisfied';

=head2 requires

requires() is a synonym for abstract() and works in the way you'd expect.

=head1 SEE ALSO

=head1 BUGS

All complex software has bugs lurking in it, and this module is no exception.

Bugs, feature requests and pull requests through GitHub are most welcome; our
page and repo (same URI):

    https://github.com/RsrchBoy/moosex-abstractmethod

=head1 AUTHOR

Chris Weyl <cweyl@alumni.drew.edu>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by Chris Weyl.

This is free software, licensed under:

  The GNU Lesser General Public License, Version 2.1, February 1999

=cut


__END__



Makefile.PL

use strict;
use warnings;

use 5.012;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Declare methods requirements that must be satisfied",
  "AUTHOR" => "Chris Weyl <cweyl\@alumni.drew.edu>",
  "BUILD_REQUIRES" => {
    "File::Find" => 0,
    "File::Temp" => 0,
    "MooseX::Traits" => 0,
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.88",
    "constant" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooseX-AbstractMethod",
  "EXE_FILES" => [],
  "LICENSE" => "lgpl",
  "NAME" => "MooseX::AbstractMethod",
  "PREREQ_PM" => {
    "Moose" => "0.94",
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "Moose::Util" => 0,
    "Moose::Util::MetaRole" => 0,
    "namespace::autoclean" => 0
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



